File: solutions/typescript/0077.ts
/**
 * 77. 组合
 * 给定两个整数 n 和 k，返回范围 [1, n] 中所有可能的 k 个数的组合。
 */
function combine(n: number, k: number): number[][] {
    const result: number[][] = [];
    const path: number[] = [];

    const dfs = (start: number): void => {
        if (path.length === k) {
            result.push([...path]);
            return;
        }
        // 剪枝: i 最大值 = n - (k - path.length) + 1，否则剩余元素不足
        const upper = n - (k - path.length) + 1;
        for (let i = start; i <= upper; i++) {
            path.push(i);
            dfs(i + 1);
            path.pop();
        }
    };

    dfs(1);
    return result;
}

export default combine;

File: solutions/rust/0077.rs
/// 77. 组合
/// 回溯 + 剪枝，返回 [1, n] 中所有 k 个数的组合
pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
    let mut result: Vec<Vec<i32>> = Vec::new();
    let mut path: Vec<i32> = Vec::with_capacity(k as usize);

    fn dfs(start: i32, n: i32, k: i32, path: &mut Vec<i32>, result: &mut Vec<Vec<i32>>) {
        if path.len() as i32 == k {
            result.push(path.clone());
            return;
        }
        let upper = n - (k - path.len() as i32) + 1;
        for i in start..=upper {
            path.push(i);
            dfs(i + 1, n, k, path, result);
            path.pop();
        }
    }

    dfs(1, n, k, &mut path, &mut result);
    result
}

impl Solution {
    pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
        combine(n, k)
    }
}
